CAMEL-11532 Java DSL enhancement to accept supplier and function arguments

sermojohn opened the pull request:
Enhanced setBody DSL with new methods accepting supplier and function argument, to align with Java 8 APIs.

davsclaus commented on camel-core/src/main/java/org/apache/camel/builder/ExpressionClause.java:
```diff
@@ -212,6 +235,17 @@ public Object evaluate(Exchange exchange) {
+    /**
+     * A functional expression of an outbound message body
+     */
+    public T outBody(final Supplier<Object> supplier) {
```
We dont really use outBody etc, so I think this can be removed

davsclaus commented:
We would need unit tests as well.

sermojohn commented:
Attached the missing unit tests and removed the redundancy.

davsclaus commented:
I dont yet see the value of the inMessage or the test is somewhat wrong, as you set a string as in message which would be wrong imho, the inMessage should be `Message` and not the message body. 

Also your unit tests need license headers.

And check the source with

    mvn compile -P sourcecheck

for code formatting problems.

sermojohn commented:
@davsclaus thank you for the review. I fixed the above issues.

I realized that `transform().inMessage(...)` replaces the OUT/IN Message body, exactly as `transform().body(...)` does. So, I removed the `inMessage(Supplier<>)` method (that was added in my first commit), as it would have the save effect as `body(Supplier<>).` It can be used when we need to calculate the new body taking the IN message into account.

Additionally, it seems (to me) that `transform(...)` does not differ from `setBody(...)` expression (also verified by comparing TransformProcessor and SetBodyProcessor). Both replace the body of IN or OUT message. What is the reason of providing both expressions (apart from backwards compatibility) ? Maybe I am missing something.

davsclaus commented:
Its EIP pattern name - setBody is not an EIP pattern name, but they do the same.

davsclaus commented:
@lburgazzoli do you mind taking a look at this

lburgazzoli commented on camel-core/src/main/java/org/apache/camel/model/ProcessorDefinition.java:
```diff
@@ -3051,6 +3052,42 @@ public Type setBody(Expression expression) {
+     * @param supplier   the supplier that provides a value to the IN message body
+     * @return the builder
+     */
+    public Type setBody(Supplier supplier) {
```
I guess a generic is missing here

sermojohn commented:
@lburgazzoli, thank you, hope it is good now. I can see that in any case the ExpressionSupport.evaluate(Exchange exchange) has Object return type, so I thought that no generics are used in the expression result type.

lburgazzoli commented:
@sermojohn yah it was more to avoid  warnings

lburgazzoli commented:
@sebbrousse could you please open the pr again to the master branch ? 
I didn't spot that it was against branch 2.20.x

sermojohn commented:
@lburgazzoli , @sebbrousse I can send a new PR to the master.
When I started this branch, the master had some build issues, so I started from the latest release branch with no build issues.
